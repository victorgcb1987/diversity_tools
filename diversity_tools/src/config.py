EXECUTABLES_REQUIREMENTS = {"meryl": {"executable": "meryl",
                                      "user_path": "MERYL_PATH"}}

MAGIC_NUMS_COMPRESSED = [b'\x1f\x8b\x08', b'\x42\x5a\x68', 
                         b'\x50\x4b\x03\x04']