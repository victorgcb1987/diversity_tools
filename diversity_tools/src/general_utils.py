
def check_results(program, results):
    if results["return_code"] == 0:
        print("{} successfully run".format(program))
    else:
        print("{} failed".format(program))
        raise RuntimeError(results["log_messages"])
    