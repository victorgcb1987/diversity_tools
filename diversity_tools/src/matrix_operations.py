import pandas as pd
import numpy as np

def convert_list_to_numbers(gene_families):
    gene_families_counts = {}
    for family_names, species in gene_families.items():
            for species_names, gene_names in species.items():
                if family_names not in gene_families_counts:
                    gene_families_counts[family_names] = {}
                if species_names not in gene_families_counts[family_names]:
                    gene_families_counts[family_names][species_names] = len(gene_names)
    return gene_families_counts

def convert_into_dataframe(gene_families):
    df = pd.DataFrame(gene_families)
    df = df.fillna(value=0)
    df = df.astype(int)
    #df = df.T
    return df

def get_dataframe_with_limited_families(df, family_number = 3):
    family_number = int(family_number)
    limited_df = df.iloc[:, :family_number]
    limited_df = limited_df.fillna(value=0)
    limited_df = limited_df.astype(int)
    return limited_df

def calculate_shannon_diversity_index(df_matrix):
    cols = list(df_matrix.columns)
    df_frequencies = pd.DataFrame(df_matrix[cols].div(df_matrix[cols].sum(axis=0), axis=1))
    df_diversity = pd.DataFrame(df_frequencies[cols].transform(lambda x: -(x*np.log(x))))
    df_shannon = df_diversity.sum(axis=0)
    return df_shannon

def filter_dataframe_cols_by_value_occurrence(df_matrix, value=1, ignore_zeros=True, threshold=1, mode = "equal"):
    df = df_matrix.fillna(value=0)
    df = df.astype(int)
    if ignore_zeros:
        threshold_cutoff = threshold * (df != 0).sum()
    else:
        threshold_cutoff = threshold * df.sum()
    if mode == "greater_than":
        df = df.loc[:, (df >= value).sum() >= threshold_cutoff]
    elif mode == "equal":
        df = df.loc[:, (df == value).sum() >= threshold_cutoff]
    elif mode == "less_than":
        df = df.loc[:, (df <= value).sum() <= threshold_cutoff]
    return df

def select_families_with_highest_number_of_genes(df_matrix, num_values):
    selected_df = df_matrix.apply(lambda x: x.nlargest(num_values, keep = 'all'))
    selected_df = selected_df.fillna(value=0)
    selected_df = selected_df.astype(int)
    return selected_df

def filter_dataframe_by_cols_name(df_matrix, column_name_list, keep_columns=True):
    df = df_matrix.fillna(value=0)
    df = df.astype(int)
    if keep_columns:
        df = df.loc[:, column_name_list]
    else:
        df = df.drop(column_name_list, axis = 1)
    return df

def filter_dataframe_by_rows_name(df_matrix, row_name_list, keep_row=True):
    df = df_matrix.fillna(value=0)
    df = df.astype(int)
    if keep_row:
        df = df.loc[row_name_list]
    else:
        df = df.drop(row_name_list, axis = 0)
    return df