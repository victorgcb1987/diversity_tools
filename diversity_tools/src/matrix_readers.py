import csv
import pandas as pd

def read_orthovenn2_composition_output(orthovenn2_input_fhand):
    orthovenn2_matrix = {}
    for line in orthovenn2_input_fhand:
        line = line.rstrip()
        line = line.split()
        family_name = line[0]
        species_gene_names = line[1]
        for species_gene_name in species_gene_names.split(','):
            species_name, gene_name = species_gene_name.split('|')
            if family_name not in orthovenn2_matrix:
                orthovenn2_matrix[family_name] = {}
            if species_name not in orthovenn2_matrix[family_name]:
                orthovenn2_matrix[family_name][species_name] = [gene_name]
            else:
                orthovenn2_matrix[family_name][species_name].append(gene_name)
    return orthovenn2_matrix


def read_matrix_from_file(gene_families_orthovenn2_Read_input_fhand, family_field = "CLNAME"):
    families_matrix = {}
    for line in csv.DictReader(gene_families_orthovenn2_Read_input_fhand):
        family_name = line.pop(family_field)
        families_matrix[family_name] = line
        for species_names, gene_count in line.items():
                    families_matrix[family_name][species_names] = float(gene_count)
    return families_matrix

def merge_inputs(rm_input, te_input):
    """Merges inputs from both RepeatMasker and TESorter.
    
    Repeats of RepeatMasker that are not in TESorter
    are filled with columns of the latter.

    Parameters
    ----------
    rm_input : list of dictionaries
        List of dictionaries from `read_repeatmasker_out`.
    
    te_input : list of dictionaries 
        List of dictionaries from `read_tesorter_cls_tsv`.

    Returns
    -------
    merged_inputs : `pandas.DataFrame`
    """
    #Create and concatenate dataframes
    rm_df = pd.DataFrame(rm_input)
    te_df = pd.DataFrame(te_input)
    concat_df = pd.concat([rm_df, te_df])

    #Fill empty values, domains contain a list
    tr = concat_df["domains"].isna()
    concat_df.loc[tr, "domains"] = concat_df.loc[tr, "domains"].apply(lambda x: [{"none":"none"}])
    first_cols = list(rm_df[rm_df.columns[~rm_df.columns.isin(['start','end', 'repeat'])]])
    last_cols = list(te_df[te_df.columns[~te_df.columns.isin(['start','end', 'repeat'])]])
    none_cols = [col for col in last_cols if col != "domains"]
    concat_df[none_cols] = concat_df[none_cols].fillna("none")

    #Merge rows of RM and TES belonging to the same repeat
    cols = first_cols + last_cols
    cols_order = {col:("first" if col in first_cols else "last") for col in cols} 
    merged_df = concat_df.groupby(["start", "end", "repeat"], as_index=False).aggregate(cols_order).reindex(columns=concat_df.columns)
    merged_df["id"] = merged_df["id"].astype("int32")
    merged_df = merged_df.sort_values(by=["id"]).reset_index(drop=True)

    #Convert values of numerical columns into floats/integers
    convert_dict = {
        "sw": "int32", "per div": "float32", "per del": "float32",
        "per ins": "float32","start": "int64", "end": "int64",
        "q left": "int64", "r start": "int32", "r end": "int32",
        "r left": "int32", "id": "int32", "length": "int32"
        }
    merged_df = merged_df.astype(convert_dict)

    return merged_df

def read_repeatmasker_out(input_fhand):
    """Reads the .out file from RepeatMasker.
    
    Each repeat is added as a dictionary, containing the data
    from each column as key-value pairs, to a common list for
    the file. Additionally, it separates the class/family
    column into two separate columns, and reallocates the columns
    of the start and (left) of "position in repeat" of repeats
    whose match is in the complementary strand.

    Parameters
    ----------
    input_fhand : file
        .out file from RepeatMasker.

    Returns
    -------
    rm_input : list of dictionaries
    """
    fieldnames = [
        "sw", "per div", "per del", "per ins", "seqid", "start",
        "end", "q left", "match", "repeat", "class/family", "r start",
        "r end", "r left", "id"
        ]
    read_repeats = []

    #Skip first three lines
    for _ in range(3):
        next(input_fhand)

    for line in input_fhand:
        repeat_data = {}
        line = line.strip(" ").split()

        for i in range(15):
            if i == 10:
                class_family = line[i].split("/")
                if len(class_family) == 1:
                    class_family.append("Unknown")
                cls = class_family[0]
                superfamily = class_family[1]
                repeat_data["class"] = cls
                repeat_data["superfamily"] = superfamily
            else:
                repeat_data[fieldnames[i]] = line[i]

        #Get right data if repeat match was in the complementary strand
        if repeat_data["match"] == "C":
            true_left = repeat_data["r start"]
            true_start = repeat_data["r left"]
            repeat_data["r start"] = true_start
            repeat_data["r left"] = true_left

        repeat_data["q left"] = repeat_data["q left"].strip("()")
        repeat_data["r left"] = repeat_data["r left"].strip("()")
        repeat_data["length"] = int(repeat_data["end"]) - int(repeat_data["start"])

        read_repeats.append(repeat_data)

    return read_repeats

def read_tesorter_cls_tsv(input_fhand):
    """Reads the .cls.tsv file from TESorter.
    
    Sequences that were analyzed by TESorter must be 
    previously analyzed by RepeatMasker (and have to be
    prepared with RepeatModeler). It creates a list
    of dictionaries, each dictionary corresponding to a repeat.
    It also reads data contained in the #TE column, which
    comes from RepeatMasker.

    Parameters
    ----------
    input_fhand : file
        .cls.tsv file from TESorter.

    Returns
    -------
    te_input : list of dictionaries
    """
    fieldnames = [
        "seqid", "start", "end", "repeat", "class",
        "superfamily", "tes order", "tes superfamily",
        "clade", "complete", "strand", "domains", "length"
        ]
    read_repeats = []
    next(input_fhand)

    for line in input_fhand:
        line = line.strip("\n").split("\t")

        #Classify data from the #TE column
        seqid_and_info = line[0].split(":", 1)
        seqid = seqid_and_info[0]
        info = seqid_and_info[1].split("_", 1)
        pos = info[0].split("..")
        start = pos[0]
        end = pos[1]
        length = int(end) - int(start)
        rep_info = info[1].split("#")
        rep = rep_info[0]

        class_family = rep_info[1].split("/")
        if len(class_family) == 1:
            class_family.append("Unknown")
        cls = class_family[0]
        family = class_family[1]

        #Get data from the other columns of TESorter
        tes_order = line[1]
        tes_superfamily = line[2]
        clade = line[3]
        complete = line[4]
        strand = line[5]

        #Data in the domains column will consist
        #of a list of dictionaries
        old_domains = line[6].split()
        new_domains = []
        for domain in old_domains:
            domain_clade = domain.split("|")
            if len(domain_clade) == 1:
                domain_clade.append("none")
            new_domains.append({domain_clade[0]: domain_clade[1]})

        #Create the dictionary for the repeat
        rep_data = [seqid, start, end, rep, cls,
                    family, tes_order, tes_superfamily,
                    clade, complete, strand, new_domains, length]
        repeat = dict(zip(fieldnames, rep_data))

        read_repeats.append(repeat)

    return read_repeats
