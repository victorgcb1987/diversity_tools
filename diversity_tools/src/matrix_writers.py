import pandas as pd

def write_file_from_matrix(families_matrix, fpath):
    df = pd.DataFrame(families_matrix)
    df = df.fillna(value=0)
    df = df.astype(int)
    df = df.T
    df.to_csv(fpath, index_label="#ID")

def write_csv_from_matrix(families_matrix, fpath):
    families_matrix.to_csv(fpath, index_label="#ID")