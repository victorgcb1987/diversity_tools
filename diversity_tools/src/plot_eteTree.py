#Made by Francois Serra (@fransua in GitHub)
#https://gist.github.com/fransua/da703c3d2ba121903c0de5e976838b71

import builtins as __builtin__
from itertools import chain
from math import floor

import matplotlib.pyplot as plt
import numpy as np

from ete3 import Tree
from matplotlib.collections import LineCollection

def round_sig(x, sig=2):
    return round(x, sig - int(floor(np.log10(abs(x)))) - 1)


def to_coord(x, y, xmin, xmax, ymin, ymax, plt_xmin, plt_ymin, plt_width, plt_height):
    x = (x - xmin) / (xmax - xmin) * plt_width  + plt_xmin
    y = (y - ymin) / (ymax - ymin) * plt_height + plt_ymin
    return x, y


def plot_tree(tree, align_names=False, name_offset=None, max_dist=None, font_size=9, axe=None, **kwargs):
    """
    Plots a ete3.Tree object using matploltib.
    
    :param tree: ete Tree object
    :param False align_names: if True names will be aligned vertically
    :param None max_dist: if defined any branch longer than the given value will be 
       reduced by this same value.
    :param None name_offset: offset relative to tips to write leaf_names. In bL scale
    :param 12 font_size: to write text
    :param None axe: a matploltib.Axe object on which the tree will be painted.
    :param kwargs: for tree edge drawing (matplotlib LineCollection) 
    :param 1 ms: marker size for tree nodes (relative to number of nodes)
    
    :returns: a dictionary of node objects with their coordinates
    """
    
    if axe is None:
        axe = plt.subplot(111)

    
    def __draw_edge_nm(c, x):
        h = node_pos[c]
        hlinec.append(((x, h), (x + c.dist, h)))
        hlines.append(cstyle)
        return (x + c.dist, h)

    def __draw_edge_md(c, x):
        h = node_pos[c]
        if c in cut_edge:
            offset = max_x / 600.
            hlinec.append(((x, h), (x + c.dist / 2 - offset, h)))
            hlines.append(cstyle)
            hlinec.append(((x + c.dist / 2 + offset, h), (x + c.dist, h)))
            hlines.append(cstyle)
            hlinec.append(((x + c.dist / 2, h - 0.05), (x + c.dist / 2 - 2 * offset, h + 0.05)))
            hlines.append(cstyle)
            hlinec.append(((x + c.dist / 2 + 2 * offset, h - 0.05), (x + c.dist / 2, h + 0.05)))
            hlines.append(cstyle)
            axe.text(x + c.dist / 2, h - 0.07, '+%g' % max_dist, va='top', 
                     ha='center', size=2. * font_size / 3)
        else:
            hlinec.append(((x, h), (x + c.dist, h)))
            hlines.append(cstyle)
        return (x + c.dist, h)

    __draw_edge = __draw_edge_nm if max_dist is None else __draw_edge_md
    
    vlinec = []
    vlines = []
    hlinec = []
    hlines = []
    nodes = []
    nodex = []
    nodey = []
    ali_lines = []
    
    # to align leaf names
    tree = tree.copy()
    max_x = max(n.get_distance(tree) for n in tree.iter_leaves())

    coords = {}
    node_pos = dict((n2, i) for i, n2 in enumerate(tree.get_leaves()[::-1]))
    node_list = tree.iter_descendants(strategy='postorder')
    node_list = chain(node_list, [tree])

    # reduce branch length
    cut_edge = set()
    if max_dist is not None:
        for n in tree.iter_descendants():
            if n.dist > max_dist:
                n.dist -= max_dist
                cut_edge.add(n)

    if name_offset is None:
        name_offset = max_x / 100.
    # draw tree
    for n in node_list:
        style = n._get_style()
        x = __builtin__.sum(n2.dist for n2 in n.iter_ancestors()) + n.dist
        if n.is_leaf():
            y = node_pos[n]
            if align_names:
                axe.text(max_x + name_offset, y, n.name, 
                         va='center', size=font_size)
                ali_lines.append(((x, y), (max_x + name_offset, y)))
            else:
                axe.text(x + name_offset, y, n.name,
                         va='center', size=font_size)
        else:
            y = np.mean([node_pos[n2] for n2 in n.children])
            node_pos[n] = y

            # draw vertical line
            vlinec.append(((x, node_pos[n.children[0]]), (x, node_pos[n.children[-1]])))
            vlines.append(style)

            # draw horizontal lines
            for child in n.children:
                cstyle = child._get_style()
                coords[child] = __draw_edge(child, x)
        nodes.append(style)
        nodex.append(x)
        nodey.append(y)

    # draw root
    __draw_edge(tree, 0)

    lstyles = ['-', '--', ':']
    hline_col = LineCollection(hlinec, colors=[l['hz_line_color'] for l in hlines], 
                              linestyle=[lstyles[l['hz_line_type']] for l in hlines],
                              linewidth=[(l['hz_line_width'] + 1.) / 2 for l in hlines])
    vline_col = LineCollection(vlinec, colors=[l['vt_line_color'] for l in vlines], 
                              linestyle=[lstyles[l['vt_line_type']] for l in vlines],
                              linewidth=[(l['vt_line_width'] + 1.) / 2 for l in vlines])
    ali_line_col = LineCollection(ali_lines, colors='k')

    axe.add_collection(hline_col)
    axe.add_collection(vline_col)
    axe.add_collection(ali_line_col)

    nshapes = dict((('circle', 'o'), ('square', 's'), ('sphere', 'o')))
    shapes = set(n['shape'] for n in nodes)
    for shape in shapes:
        indexes = [i for i, n in enumerate(nodes) if n['shape'] == shape]
        scat = axe.scatter([nodex[i] for i in indexes], 
                           [nodey[i] for i in indexes], 
                           s=0, marker=nshapes.get(shape, shape))
        scat.set_sizes([(nodes[i]['size'])**2 / 2 for i in indexes])
        scat.set_color([nodes[i]['fgcolor'] for i in indexes])
        scat.set_zorder(10)

    # scale line
    xmin, xmax = axe.get_xlim()
    ymin, ymax = axe.get_ylim()
    diffy = ymax - ymin
    dist = round_sig((xmax - xmin) / 5, sig=1)
    ymin -= diffy / 100.
    axe.plot([xmin, xmin + dist], [ymin, ymin], color='k')
    axe.plot([xmin, xmin], [ymin - diffy / 200., ymin + diffy / 200.], color='k')
    axe.plot([xmin + dist, xmin + dist], [ymin - diffy / 200., ymin + diffy / 200.], 
             color='k')
    axe.text((xmin + xmin + dist) / 2, ymin - diffy / 200., dist, va='top',
             ha='center', size=font_size)
    axe.set_axis_off()
    return coords