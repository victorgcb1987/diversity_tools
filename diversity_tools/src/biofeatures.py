from src.dependencies import get_executables
from src.config import EXECUTABLES_REQUIREMENTS as exec_reqs
from src.general_utils import check_results, store_results
from subprocess import run


def extract_biofeatures(gff_fpath, genome_fpath, out_fpath, kind=None):
    if kind is None:
        msg = "Feature not recognised. Available features: TE, gene, intron"
        raise ValueError(msg)
    elif kind == "TE":
        feature_seqs_fpath = write_sequences_from_bed(gff_fpath, genome_fpath, out_fpath)
    elif kind == "gene":
        feature_seqs_fpath = write_gene_sequences(gff_fpath, genome_fpath, out_fpath)
    elif kind == "intron":
        feature_seqs_fpath = write_intron_sequences(gff_fpath, genome_fpath, out_fpath)


def write_sequences_from_bed(bed_fpath, genome_fpath, out_fpath, kind="TE"):
    bedtools_executable = get_executables(exec_reqs["bedtools"])
    out_fpath = out_fpath / "{}.{}s.fa".format(genome_fpath.stem, kind)
    cmd = [bedtools_executable, "getfasta", "-fi", str(genome_fpath), 
           "-bed", str(bed_fpath), "-s", ">", str(out_fpath)]
    print(cmd[5])
    run_bedtools = run(" ".join(cmd), shell=True, capture_output=True)
    results = store_results(out_fpath, run_bedtools)
    check_results("bedtools, extracting {}s.".format(kind), results)   
    return results


def write_genome_size(genome_fpath, out_fpath):
    #Index the genome
    samtools_executable = get_executables(exec_reqs["samtools"])
    cmd = [samtools_executable, "faidx", str(genome_fpath)]
    faidx_fpath = genome_fpath.parent / "{}.fai".format(genome_fpath.name)
    run_faidx = run(" ".join(cmd), shell=True, capture_output=True)
    results_faidx = store_results(faidx_fpath, run_faidx)
    check_results("samtools, creating genome index", results_faidx)
    
    #Generate file with start-end of pseudomolecules
    chromsizes_fpath = out_fpath / "{}_chromsizes.bed".format(genome_fpath.stem)
    cmd = ["awk", "\'OFS=\"\t\" {print $1, \"1\", $2}\'",
           str(faidx_fpath),  "| sort -k1,1 -k2,2n > ",
           str(chromsizes_fpath)]
    run_chromsizes = run(" ".join(cmd), shell=True, capture_output=True)
    results_chromsizes = store_results(chromsizes_fpath, run_chromsizes)
    check_results("AWK, creating chromsizes", results_chromsizes)
    return results_chromsizes

def sort_gff(gff_fpath, out_fpath):
    sorted_gff_fpath = out_fpath / "{}.sorted.gff".format(gff_fpath.stem)
    cmd = ["cat", str(gff_fpath), "|",
           "awk \'$1 ~ /^#/ {print $0;next} {print $0 | \"sort -k1,1 -k4,4n -k5,5n\"}\'",
           ">", str(sorted_gff_fpath)]
    run_sort_gff = run(" ".join(cmd), shell=True, capture_output=True)
    results_sort_gff = store_results(sorted_gff_fpath, run_sort_gff)
    check_results("AWK, sorting gff", results_sort_gff)
    return results_sort_gff


def get_intergeninc_regions(gff_fpath, genome_fpath, chromsizes_fpath, out_fpath):
    intergenic_bed_fpath = out_fpath / "{}_intergenic.bed".format(genome_fpath.stem)
    bedtools_executable = get_executables(exec_reqs["bedtools"])
    cmd = [bedtools_executable, "complement", "-i",
           str(gff_fpath),  "-g", str(chromsizes_fpath),
           ">", str(intergenic_bed_fpath)]
    run_intergenic = run(" ".join(cmd), shell=True, capture_output=True)
    results_intergenic = store_results(intergenic_bed_fpath, run_intergenic)
    check_results("Bedtools, creating intergenic regions", results_intergenic)
    return results_intergenic


def write_gene_sequences(gff_fpath, genome_fpath, out_fpath):
    gffread_executable = get_executables(exec_reqs["gffread"])
    out_fpath = out_fpath / "{}.genes.fa".format(out_fpath.parent)
    cmd = [gffread_executable, "-g", str(genome_fpath), "-w", str(out_fpath), str(gff_fpath)]
    run_gffread = run(" ".join(cmd), shell=True, capture_output=True)
    results = store_results(out_fpath, run_gffread)
    check_results("gffread, extracting genes", results)   
    return results


def write_exon_regions(gff_fpath, genome_fpath, out_fpath):
    exons_fpath = out_fpath / "{}_exons.bed".format(genome_fpath.stem)
    with open(gff_fpath) as input_fhand:
        with open(exons_fpath, "w") as out_fhand:
            for line in input_fhand:
                if line.startswith("#") or not line:
                    continue
                line = line.rstrip().split()
                if line[2] == "exon":
                    out_fhand.write("{}\t{}\t{}\n".format(line[0], line[3], line[4]))
                    out_fhand.flush()
    results = {"out_fpath": exons_fpath,
               "return_code": 0,
               "log_messages": "OK"}
    check_results("Creating exons bed file", results)
    return results


def write_intron_regions(intergenic_regions, exon_regions, chromsizes_fpath, genome_fpath, out_fpath):
    #concatenate intergenic and exon bed files
    concatenated_regions_fpath = out_fpath / "{}_intergenic_and_exonic_regions.bed".format(genome_fpath.stem)
    cmd = ["cat", str(exon_regions), str(intergenic_regions), "|", "sort -k1,1 -k2,2n",
           ">", str(concatenated_regions_fpath)]
    run_concatenate = run(" ".join(cmd), shell=True, capture_output=True)
    concatenate_results = store_results(concatenated_regions_fpath, run_concatenate)
    check_results("Concatenate regions", concatenate_results)

    intron_regions_fpath = out_fpath / "{}_intragenic_regions.bed".format(genome_fpath.stem)
    bedtools_executable = get_executables(exec_reqs["bedtools"])
    cmd = [bedtools_executable, "complement", "-i", 
           "{}".format(str(concatenated_regions_fpath)), 
           "-g", str(chromsizes_fpath), ">", str(intron_regions_fpath)]
    run_intragenic_bed = run(" ".join(cmd), shell=True, capture_output=True)    
    results = store_results(intron_regions_fpath, run_intragenic_bed)
    check_results("Getting intragenic regions", results)
    return results


def write_intron_sequences(gff_fpath, genome_fpath, out_fpath):
    chromsizes_fpath = write_genome_size(genome_fpath, out_fpath)["out_fpath"]
    sorted_gff_path = sort_gff(gff_fpath, out_fpath)["out_fpath"]
    intergenic_fpath = get_intergeninc_regions(sorted_gff_path, genome_fpath, 
                                               chromsizes_fpath, out_fpath)["out_fpath"]
    exonic_fpath = write_exon_regions(gff_fpath, genome_fpath, out_fpath)["out_fpath"]
    intronic_fpath = write_intron_regions(intergenic_fpath, exonic_fpath, 
                                          chromsizes_fpath, genome_fpath, out_fpath)["out_fpath"]
    introns_sequences_fpath = write_sequences_from_bed(intronic_fpath, genome_fpath, out_fpath, kind="intron")
    return introns_sequences_fpath