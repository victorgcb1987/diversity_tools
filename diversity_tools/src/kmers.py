from src.dependencies import get_executables
from src.config import EXECUTABLES_REQUIREMENTS as exec_reqs
from subprocess import run


def run_meryl(sequences_fpath, out_fdir, threads=1, kmer_size=21):
    meryl_executable = get_executables(exec_reqs["meryl"])
    cmd = [meryl_executable, "count", "k={}".format(kmer_size), 
           str(sequences_fpath), "threads={}".format(threads), 
           "output", str(out_fdir)+"/"]
    print(" ".join(cmd))
    meryl_run = run(" ".join(cmd), shell=True, capture_output=True)
    results = {"output_fdir": out_fdir,
               "return_code": meryl_run.returncode,
               "log_messages": meryl_run.stderr.decode()}
    return results

def count_meryl_kmers(out_fdir, num_mismatches=3):
    meryl_executable = get_executables(exec_reqs["meryl"])
    cmd = [meryl_executable, "print", str(out_fdir)+"/"]
    count_meryl_run = run(" ".join(cmd), shell=True, capture_output=True)
    results = {"output_fdir": out_fdir,
               "return_code": count_meryl_run.returncode,
               "log_messages": count_meryl_run.stderr.decode()}
    kmers = count_meryl_run.stdout.decode()
    kmer_occurrency = {}
    for line in kmers:
        line = line.split()
        kmer = line[0]
        num_occurrences = int(line[1])
        kmer_occurrency[kmer] = num_occurrences
    return kmer_occurrency
