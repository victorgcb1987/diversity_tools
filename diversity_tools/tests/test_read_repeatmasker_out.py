import unittest
from pathlib import Path

import pandas as pd
from pandas.testing import assert_frame_equal

from src.matrix_readers import read_repeatmasker_out

class RepMask(unittest.TestCase):

    def setUp(self):
        test_path = Path(__file__).parent.absolute()
        test_path = test_path / "test_data"
        self.test_path = test_path / "test_read_repeatmasker_out.out"

    def test_read_repmask(self):
        with open(self.test_path) as input_fhand:
            read_repeats = read_repeatmasker_out(input_fhand)
        convert_dict = {
        "per div": "float16", "per del": "float16",
        "per ins": "float16","start": "int32", "end": "int32",
        "repeat": "category", "length": "int32", "class": "category",
        "superfamily": "category"
        }
        test_df = pd.DataFrame([{
            "per div": "14.7", "per del": "2.1", "per ins": "2.4",
            "seqid": "Peame105C00", "start": "23", "end": "4959",
            "repeat": "rnd-4_family-1935", "length": 4936, "class": "Unknown",
            "superfamily": "Unknown"
            },{
            "per div": "31.9", "per del": "5.0", "per ins": "5.0",
            "seqid": "Peame105C00", "start": "4959", "end": "5519",
            "repeat": "ltr-1_family-724", "length": 560, "class": "LTR",
            "superfamily": "Copia"
            },{
            "per div": "32.2", "per del": "4.6", "per ins": "4.6",
            "seqid": "Peame105C00", "start": "5630", "end": "5847",
            "repeat": "ltr-1_family-645",  "length": 217, "class": "LTR",
            "superfamily": "Copia"
            }]).astype(convert_dict)
        
        assert_frame_equal(test_df, read_repeats)
        
if __name__ == "__main__":
    unittest.main()
