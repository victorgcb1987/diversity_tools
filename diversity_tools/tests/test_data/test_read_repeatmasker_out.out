    SW   perc perc perc  query        position in query                 matching           repeat                position in repeat
 score   div. del. ins.  sequence     begin     end            (left)   repeat             class/family      begin   end    (left)       ID

 25785   14.7  2.1  2.4  Peame105C00         23      4959  (55849342) + rnd-4_family-1935  Unknown                 1   4924     (0)       1  
   434   31.9  5.0  5.0  Peame105C00       4959      5519  (55848782) C ltr-1_family-724   LTR/Copia          (2165)   6492    5932       2 *
   225   32.2  4.6  4.6  Peame105C00       5630      5847  (55848454) C ltr-1_family-645   LTR/Copia           (746)   3334    3117       3  
